plugins/node/instrumentation-fs/src/instrumentation.ts

```diff
@@ -0,0 +1,321 @@
+        instrumentation.getConfig() as FsInstrumentationConfig;
+      if (typeof createHook === 'function') {
+        if (
+          createHook(fnName, {
+            args: args,
+          }) === false
+        ) {
+          // return original.apply(this, arguments);
+          return api.context.with(
+            suppressTracing(api.context.active()),
+            original,
+            this,
+            ...args
+          );
+        }
+      }
+
+      const lastIdx = args.length - 1;
+      const cb = args[lastIdx];
+      if (typeof cb === 'function') {
+        const span = instrumentation.tracer.startSpan(
+          `fs ${fnName}`
+        ) as api.Span;
+
+        // return to the context active during the call in the callback
+        args[lastIdx] = api.context.bind(
+          api.context.active(),
+          function (this: unknown, err?: Error) {
+            if (err) {
+              span.recordException(err);
+              span.setStatus({
+                message: err.message,
+                code: api.SpanStatusCode.ERROR,
+              });
+            }
+            if (typeof endHook === 'function') {
+              try {
+                endHook(fnName, { args: arguments, span });
+              } catch (e) {
+                instrumentation._diag.error('caught endHook error', e);
+              }
+            }
+            span.end();
+            return cb.apply(this, arguments);
+          }
+        );
+
+        try {
+          return api.context.with(
+            api.trace.setSpan(api.context.active(), span),
+            original,
+            this,
+            ...args
+          );
+        } catch (err) {
+          span.recordException(err);
+          span.setStatus({
+            message: err.message,
+            code: api.SpanStatusCode.ERROR,
+          });
+          if (typeof endHook === 'function') {
+            try {
+              endHook(fnName, { args: args, span });
+            } catch (e) {
+              instrumentation._diag.error('caught endHook error', e);
+            }
+          }
+          span.end();
+          throw err;
+        }
+      } else {
+        // TODO: what to do if we are pretty sure it's going to throw
+        return original.apply(this, args);
+      }
+    };
+  }
+
+  protected _patchSyncFunction<T extends (...args: any[]) => ReturnType<T>>(
+    fnName: string,
+    original: T
+  ): T {
+    const instrumentation = this;
+    return <any>function (this: any, ...args: any[]) {
+      if (isTracingSuppressed(api.context.active())) {
+        // Performance optimization. Avoid creating additional contexts and spans
+        // if we already know that the tracing is being suppressed.
+        return original.apply(this, args);
+      }
+      const { createHook, endHook } =
+        instrumentation.getConfig() as FsInstrumentationConfig;
+      if (typeof createHook === 'function') {
+        if (
+          createHook(fnName, {
```

Comment:
try / catch to protect against exception?

Reply:
It's a bug here, yes. I refactored running hooks altogether and brought running those into one place and making them safe to run in one go.